### PR TITLE
Various fixes

### PR DESCRIPTION
**Pull request by WillB97**
First pass review of current codebase.

### CODE REVIEW

File: kchd/app.py [hunk 1]
```diff
@@ -39,10 +39,11 @@ def _init(self) -> None:
...
         }
 
-        leds = set().union(*[
-            controller.leds  # type: ignore[attr-defined]
+        leds = {
+            led
             for controller in self._controllers.values()
-        ])
+            for led in controller.leds  # type: ignore[attr-defined]
+        }
```
Comment (trickeydan) [inline kchd/app.py:46]: I don't think this is particularly easy to understand.

Equally, the current implementation sucks and this definitely an improvement.

File: kchd/controllers/system_status.py [hunk 2]
```diff
@@ -86,19 +85,14 @@ def astoria_good(self) -> bool:
...
-        return sum([1 for factor in led_factors if factor])
 
     def get_state(self) -> Dict[KCHLED, bool]:
         """Get the state of controlled LEDs."""
-        leds_on = self.get_enabled_leds() + 2
-        return {
-            led: leds_on >= i + 1
-            for i, led in enumerate(self.leds)
-        }
+        return dict(zip(self.leds, self.get_enabled_leds()))
```
Comment (trickeydan) [inline kchd/controllers/system_status.py:98]: In my experience people who are not used to Python find dict comprehensions easier to understand.

```suggestion
        return {led: state for led, state in self.get_enabled_leds().items()}
```

File: kchd/gpio.py [hunk 3]
```diff
@@ -15,7 +15,6 @@ def __init__(self, leds: Set[KCHLED]) -> None:
 
         atexit.register(GPIO.cleanup)
         GPIO.setmode(GPIO.BCM)
-        GPIO.setwarnings(False)  # Ignore that the BOOT_20 and BOOT_40 LEDs were in use.
```
Comment (trickeydan) [inline kchd/gpio.py:18]: We need to keep this as KCH intentionally takes over these LEDs.

When the updating animation is implemented, all of the boot LEDs will be used.

File: kchd/hardware.py [hunk 4]
```diff
@@ -10,16 +10,16 @@ class KCHLED(enum.IntEnum):
     The value of each LED in the Enum is the BCM pin number.
     """
 
-    BOOT_20 = 7
-    BOOT_40 = 5
+    # BOOT_20 = 7  # Controlled by hat EEPROM GPIO map
+    # BOOT_40 = 5  # Controlled by systemd
```
Comment (trickeydan) [inline kchd/hardware.py:14]: We need to keep this as KCH intentionally takes over these LEDs.

When the updating animation is implemented, all of the boot LEDs will be used.

File: kchd/hardware.py [hunk 5]
```diff
@@ -10,16 +10,16 @@ class KCHLED(enum.IntEnum):
...
     BOOT_60 = 12
     BOOT_80 = 6
     BOOT_100 = 13
 
     CODE = 11
     COMP = 16
 
-    HEARTBEAT = 19
+    # HEARTBEAT = 19  # Controlled by hat EEPROM device tree
```
Comment (trickeydan) [inline kchd/hardware.py:22]:
```suggestion
    HEARTBEAT = 19  # Controlled by hat EEPROM device tree
```

As the `HEARTBEAT` LED is not in use by any of the `LEDController` classes, this LED is never touched by kchd anyway. There is little advantage to commenting it out.

File: kchd/controllers/system_status.py [hunk 6]
```diff
@@ -86,19 +85,14 @@ def astoria_good(self) -> bool:
         """Determine whether astoria services are in a good state."""
         return len(self._seen_services) == len(self._required_services)
 
-    def get_enabled_leds(self) -> int:
+    def get_enabled_leds(self) -> Tuple[bool, bool, bool]:
```
Comment (trickeydan) [inline kchd/controllers/system_status.py:88]: The object returned from this function is not a tuple.

```suggestion
    def get_enabled_leds(self) -> List[bool]:
```

**Review by trickeydan — CHANGES_REQUESTED** [review]
kch needs to take over the 20% and 40% boot LEDs from the system as the LEDs will be used later in development for the update animation.

This is because we'd like to use the progress LEDs as an indicator of the update progress. (Probably flashing rather than solid)

File: kchd/hardware.py (hunk 5 above)
Comment (WillB97) [reply kchd/hardware.py:22]: > There is little advantage to commenting it out.

Quite the opposite, by removing it from the enum if someone (who may not be aware that it is being driven externally) attempts to use it they will get an error.

File: kchd/controllers/system_status.py (hunk 6 above)
Comment (WillB97) [reply kchd/controllers/system_status.py:88]: This was a mistake: a636fc5

File: kchd/hardware.py (hunk 5 above)
Comment (trickeydan) [reply kchd/hardware.py:22]: @raccube Please could we have a third opinion here?

File: kchd/app.py [hunk 7]
```diff
@@ -39,6 +39,7 @@ def _init(self) -> None:
             "status": SystemStatusController(self._mqtt, self.update_leds),
         }
 
+        # Create a flattened, unique list of all leds used by all controllers
```
Comment (trickeydan) [inline kchd/app.py:42]:
```suggestion
        # Create a flattened, unique set of all leds used by all controllers
```

**Comment by WillB97** [comment]
> kch needs to take over the 20% and 40% boot LEDs from the system as the LEDs will be used later in development for the update animation.
> 
> This is because we'd like to use the progress LEDs as an indicator of the update progress. (Probably flashing rather than solid)

I would stronly advise against using the boot progress LEDs for anything other than boot progress as it makes the output ambiguous and harder to explain in the documentation. 

If we added the constraint that usercode isn't run while updating we could use the user LEDs to show the progress with, for instance a flashing code LED or blue status LED to denote that it is updating.

**Comment by trickeydan** [comment]
> > kch needs to take over the 20% and 40% boot LEDs from the system as the LEDs will be used later in development for the update animation.
> 
> > 
> 
> > This is because we'd like to use the progress LEDs as an indicator of the update progress. (Probably flashing rather than solid)
> 
> 
> 
> I would stronly advise against using the boot progress LEDs for anything other than boot progress as it makes the output ambiguous and harder to explain in the documentation. 
> 
> 
> 
> If we added the constraint that usercode isn't run while updating we could use the user LEDs to show the progress with, for instance a flashing code LED or blue status LED to denote that it is updating.

Very well. You'll need to remove them from the LEDs list in the system status controller so that mypy doesn't complain that the attributes don't exist in the enum

**Review by trickeydan — APPROVED** [review]

**Pull request merged**